### PR TITLE
Suggested improvements/testing and plugins

### PR DESCRIPTION
**Pull request by jaketf**
Hello! I am working on improving the source for this example to be more easily extensible by our customers.

For now this PR is a suggestion to serve as a discussion piece for improvements.
I'd eventually like to merge this but would need to coordinate with an update to [this solutions article ](https://cloud.google.com/solutions/cicd-pipeline-for-data-processing) to reflect the changes.

I've noticed and have addressed  the following gaps:

- [x] Best practice to have custom operators implemented as airflow plugins stored in the plugins directory on GCS.
- [x] Best practice to define all airflow variables as version controlled code. This can be done as a json config file.
- [x] Explicitly use `python3` interpreter (Post should be updated to create a python3 environment as airflow is sun-setting `python2` support.). 
- [x] Add Dag Validation test before uploading to Composer (to catch syntax issues early)
- [x] Improve directory structure to be more inline with connonical python project structure:

```bash
tree ci-cd-for-data-processing-workflow
ci-cd-for-data-processing-workflow/
├── config
│   └── Variables.json
├── CONTRIBUTING.md
├── env-setup
│   ├── create_buckets.sh
│   ├── set_composer_variables.sh
│   └── set_env.sh
├── LICENSE
├── README.md
└── source-code
    ├── build-pipeline
    │   ├── build_deploy_test.yaml
    │   ├── deploy_prod.yaml
    │   └── wait_for_dag_deployed.sh
    ├── dags
    │   ├── prod_word_count.py
    │   ├── support-files
    │   │   ├── input.txt
    │   │   └── ref.txt
    │   ├── sql
    │   │   ├── shakespeare_top_25.sql
    │   └── test_word_count.py
    ├── data-processing-code
    │   ├── pom.xml
    │   └── src
    │       ├── main
    │       │   └── java
    │       │       └── org
    │       │           └── apache
    │       │               └── beam
    │       │                   └── examples
    │       │                       └── WordCount.java
    │       └── test
    │           └── java
    │               └── org
    │                   └── apache
    │                       └── beam
    │                           └── examples
    │                               └── WordCountTest.java
    ├── plugins
    │   └── xcom_utils_plugin
    │       ├── __init__.py
    │       └── operators
    │           └── compare_xcom_maps.py
    └── tests
        └── test_compare_xcom_maps.py

```
- [x] Use `unittest`’s test discovery feature to handle adding more tests modules to the `tests/` directory 
- [x] Best practice dag definition filename should be the same as dag id.
- [x] Add `.gitignore` for common python and airflow junk. 
- [x] Improve readability (`pylint` scores == 10.0)
- [x] `BUILD` script to source env-vars and submit the build for convenience.
- [x] Add script to create composer environment set's `dags_are_paused_at_creating=True` which improves stability of compser UI but requires dags to be unpaused after deployment.

The DAGs in the original example have no schedule interval and are paused on upload which is only realistic for externally triggered DAGs and is hiding a lot of complexity in deploying DAGs with a schedule. I will address these in a subsequent PR:
  - [x] Add a step to the deploy to unpause the DAG.
  - [x] This is not a complete Solution for CICD without 
      - [x] The ability to deploy updates to a DAG or remove a DAG.
          - [x] This should be a delete of the old dag and deploy of the new DAG to avoid undetermined behavior of updating a DAG while it is running.
      - [x] The ability to remove a DAG.
          - With previous customers I've done this by having a running_dags.txt with all the DAG id's that should be running as part of this deployment. Then you do a difference with the results of `list_dags` on the target environment.
      - [x] The ability to deploy a new DAG with out redeploying all existing DAGs.

### CODE REVIEW

**Comment by jaketf** [comment]
@kingman Kindly request your review on this.

**Comment by jaketf** [comment]
@kingman Added some more bells and whistles. This seems stable now and we can start review and thinking about how to update the solution article.

File: env-setup/create_composer_environment.sh [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+gcloud composer environments create $COMPOSER_ENV_NAME \
+    --location $COMPOSER_REGION \
+    --zone $COMPOSER_ZONE_ID \
+    --machine-type n1-standard-4 \
```
Comment (kingman) [inline env-setup/create_composer_environment.sh]: For the tutorial we want to use the small instance to minimize the cost of running.
change to `n1-standard-1`

File: config/Variables.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (kingman) [inline config/Variables.json]: This is the output of the processed variables file, we need to have original file for the envsubst to work

File: env-setup/render_variables.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+# limitations under the License.
+
+# This stages the Variables file in GCS in the data directory which
+# gets synced to /home/airflow/gcs/data/ so that an airflow command
+# can reference the file locally to the worker it is running on.
+source ./set_env.sh
+source ./get_composer_properties.sh
+
+cat ../config/Variables.json | envsubst > ../config/Variables.json.processed
+mv ../config/Variables.json.processed ../config/Variables.json
```
Comment (kingman) [inline env-setup/render_variables.sh]: use template file and not override the Variables.json file

File: env-setup/set_env.sh [hunk 4]
```diff
@@ -19,6 +19,7 @@
 # limitations under the License.
 export TEST='test'
 export GCP_PROJECT_ID=$(gcloud config list --format 'value(core.project)')
+export PROJECT_ID=$(gcloud config list --format 'value(core.project)')
```
Comment (kingman) [inline env-setup/set_env.sh]: export PROJECT_ID="${GCP_PROJECT_ID}"
No need to run the same command twice

**Pull request closed without merging**